File: cpp_solutions/LinkedList/Reorder_List.cpp
/*
* Given a singly linked list L: L0→L1→…→Ln-1→Ln,

* reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…

* You must do this in-place without altering the nodes' values.

* For example,

* Given {1,2,3,4}, reorder it to {1,4,2,3}.
*
**/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        if(!head || !head->next) return;
   		
   		// Split this list to two parts
   		ListNode* slow = head;
   		ListNode* fast = head;

   		while(fast->next && fast->next->next){
   			slow = slow->next;
   			fast = fast->next->next;
   		}     

   		fast = slow->next;
   		slow->next = NULL;

   		// reverse right part;
   		ListNode* dummy = new ListNode(0);
   		while(fast){
   			ListNode* n = dummy->next;
   			dummy->next = fast;
   			ListNode* nn = fast->next;
   			fast->next = n;
   			fast = nn;
   		}

   		slow = head;
   		fast = dummy->next;

   		// merge
   		while(slow){
   			if(fast != NULL){
   				ListNode* n = slow->next;
   				slow->next = fast;
   				ListNode* nn = fast->next;
   				fast->next = n;
   				fast = nn;
   				slow = n;
   			} else{
   				break;
   			}
   		}
    }
};     

File: python_solutions/LinkedList/Reorder_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: void Do not return anything, modify head in-place instead.
        """
        if head is None or head.next is None:
            return
        
        slow, fast = head, head
        while fast.next and fast.next.next:
            fast = fast.next.next
            slow = slow.next
        
        fast = slow.next
        slow.next = None
        
        # reverse fast
        pCur = None
        while fast != None:
            temp = fast.next
            fast.next = pCur
            pCur = fast
            fast = temp
        
        # Merge 
        tail = head
        while pCur:
            pNext = pCur.next
            pCur.next = tail.next
            tail.next = pCur
            tail = tail.next.next
            pCur = pNext
        
    